### ISSUE TITLE
The latest July 5 snapshot of the Swift 5.7 compiler cannot build the NIO tests on linux/Android

### ISSUE THREAD

**Issue by buttaface** [apple/swift#59913]
### Expected behavior
All tests to build and pass.

### Actual behavior
I get this compilation error instead:
```
swift-nio/Tests/NIOPosixTests/EventLoopTest.swift:506:41: error: mutation of captured var 'promises' in concurrently-executing code
                 promiseQueue.sync { promises.append(promise) }
```
### Steps to reproduce

1. Get and install latest 5.7 official snapshot build.
2. `swift test`

### If possible, minimal yet complete reproducer code (or URL to code)

I ~did~do not see this error with the previous June 26 5.7 snapshot, ~though I did not try that~even after trying it with the latest NIO commit, implying something changed in the 5.7 compiler in the last week to cause this.

### SwiftNIO version/commit hash

apple/swift-nio@f1d62128c

### System & version information

Ubuntu 20.04 x86_64 and Swift 5.7 July 5 snapshot, both compiling natively and [cross-compiling for Android](https://github.com/buttaface/swift-android-sdk/runs/7211545158?check_suite_focus=true#step:13:1200)

**Comment by Lukasa** [comment]
It's not clear to me that the compiler is wrong here. We are absolutely mutating an Array from multiple threads. The compiler doesn't understand locks, so `DispatchQueue.sync` cannot prevent the compiler from moaning about this: Sendable just doesn't operate at that scale.

**Comment by buttaface** [comment]
I have not looked at the NIO code: I filed this issue under NIO initially because I was unsure if it was a compiler regression or something in NIO it was flagging correctly, as I don't use Swift concurrency.

**Comment by Lukasa** [comment]
Oh, this file hasn’t adopted concurrency: this sounds like a bug then.

**Comment by dnadoba** [comment]
I was able to create a reduced example:
```swift
// ModuleA
@preconcurrency
public func preconcurrencySendableClosure(_ callback: @escaping @Sendable () -> Void) {}
```
```swift
// ModuleB
import ModuleA
import Foundation

func foo() {
    let dispatchQueue = DispatchQueue(label: "SendableError")
    var mutableVar = 1
    preconcurrencySendableClosure {
        dispatchQueue.sync {
            mutableVar += 1
        }
    }
}
```
It produces the same error if all code is in the same module.
Most notably is that if I replace the call to `DispatchQeueu.sync(_:)` with a call to a function with the same type signature e.g.:
```swift
// ModuleA
public func nonEscaping(_ callback: () -> ()) {}
```
 it doesn't produce a warning.
I guess something is special about `DispatchQueue.sync(_:)` but I don't know what.

**Comment by kavon** [comment]
A `@preconcurrency` function with a `@Sendable` argument type doesn't make sense, and seems like should be rejected or warned about, since I can imagine `@preconcurrency` has really no effect anymore. Is that right @beccadax 

Also, DispatchQueue.sync (and other similar Dispatch functions) are [specially recognized by the compiler](https://github.com/apple/swift/blob/c8f4b09809de1fab3301c0cfc483986aa6bdecfa/lib/Sema/TypeCheckConcurrency.cpp#L4588) to be treated as if its parameter were annotated with `@_unsafeSendable`, which will only be treated as `@Sendable` if the argument is passed from a concurrency-adopting context. Since `sync` is being called within a `@Sendable` closure, the error about the concurrent mutation looks correct to me here :

```
var mutableVar = 1
    preconcurrencySendableClosure {
        dispatchQueue.sync {
            mutableVar += 1 // error mutation of captured var 'mutableVar' in concurrently-executing code
        }
    }
```

**Comment by Lukasa** [comment]
> A @preconcurrency function with a @Sendable argument type doesn't make sense to me, and seems like should be rejected or warned about, since I can imagine @preconcurrency has really no effect anymore.

The intention is to signal “this closure argument should always have been `@Sendable`, but it was introduced before concurrency, and the caller may not have adopted Sendable checking yet, so only warn about misuse if they have”.

**Comment by kavon** [comment]
Gotcha. Since `@_unsafeSendable` is gone, I guess that is the only way to do it.

**Comment by dnadoba** [comment]
Fixed in the latest 5.7 toolchain from July 17, 2022
